### PR TITLE
Install `fastapi`, `pydantic` and `orjson` during extension activation & bundle kedro-viz into extension

### PR DESCRIPTION
**Pull request by jitu5**
Resolves https://github.com/kedro-org/vscode-kedro/issues/72

As `pydantic` and `orjson` depends on platform and python version, So it cant be packaged with extension so those dependencies are installed during extension activation in Kedro extension env (bundled/libs)

### CODE REVIEW

File: bundled/tool/install_dependencies.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    except ImportError:
+        subprocess.check_call(
+            [
+                sys.executable,
+                "-m",
+                "pip",
+                "install",
+                "fastapi>=0.100.0,<0.200.0",
+                "pydantic>=2.0.0", # In case of FastAPI installs pydantic==1
+                "orjson>=3.9, <4.0",
```
Comment (noklam) [inline bundled/tool/install_dependencies.py:18]: NIL: can we move this to something like a requirement file, I just think it's a maintainence burden to have dependencies scatter in both code and file.

File: noxfile.py [hunk 2]
```diff
@@ -24,6 +24,18 @@ def _install_bundle(session: nox.Session) -> None:
...
+    # Installing kedro-viz to bundled/libs from here and not keeping it in requirements.in to avoid installing all its required packages
+    session.install(
+        "-t",
+        "./bundled/libs",
+        "--no-cache-dir",
+        "--implementation",
+        "py",
+        "--no-deps",
+        "--upgrade",
+        "kedro-viz",
```
Comment (noklam) [inline noxfile.py:36]: For viz, do we have a particular version that we need to pin? (I think it's the next release version, so we need to wait until it is released, but I just want to confirm this).

File: src/extension.ts [hunk 3]
```diff
@@ -28,15 +28,29 @@ import { loadServerDefaults } from './common/setup';
...
 
 let lsClient: LanguageClient | undefined;
 
 export async function getlsClient() {
     return lsClient;
 }
 
-
 export async function activate(context: vscode.ExtensionContext): Promise<void> {
+    const alreadyInstalled = context.globalState.get('dependenciesInstalled', false);
```
Comment (noklam) [inline src/extension.ts:40]: What is this globalState? Does that mean it only install if the dependency is not there? I couldn't find the code where this state is set.

File: bundled/tool/install_dependencies.py (hunk 1 above)
Comment (jitu5) [reply bundled/tool/install_dependencies.py]: done.

File: noxfile.py (hunk 2 above)
Comment (jitu5) [reply noxfile.py:36]: Mostly it would be next release `10.0.0`. later we willl pin it

File: src/extension.ts (hunk 3 above)
Comment (jitu5) [reply src/extension.ts:40]: In a VSCode extension, `globalState` is an API provided by the VSCode framework that allows you to store and retrieve global state info. This state is persisted across VS Code sessions, it remains available even after VS Code is closed and reopened. 

`globalState` does not have `.set` method.
Screenshot 2024-08-28 at 6 30 37 p m

> Does that mean it only install if the dependency is not there? 

No it just check if `await installPythonDependencies(context);` is called before with `alreadyInstalled` flag.

**Comment by jitu5** [comment]
@noklam you mentioned in slack 
> On client side (TS), we have a function that is callPythonScript
> On Python side we keep install_dependencies as a executable script.

Could you please elaborate more this, so that I can make changes accordingly.

**Comment by noklam** [comment]
I think it's mainly just a refactoring, you have this now:

```ts
export async function installPythonDependencies(context: any): Promise<void> {
    return new Promise<void>(async (resolve, reject) => {
        const script = context.asAbsolutePath('bundled/tool/install_dependencies.py');
        const interpreterDetails = await getInterpreterDetails();
        const pythonPath = interpreterDetails['path'] && interpreterDetails['path'][0];

        cp.exec(`${pythonPath} ${script} ${EXTENSION_ROOT_DIR}`, (error, stdout, stderr) => {
            if (error) {
                console.error(stderr);
                reject(error);
            } else {
                console.log(stdout);
                resolve();
            }
        });
    });
}
```

I was thinking something more like this to make it slightly generic. Frankly it is fine to keep it as is as well.
```ts
export async function callPythonScript(context: any, path_to_script: str): Promise<void> {
    return new Promise<void>(async (resolve, reject) => {
        const script = context.asAbsolutePath(path_to_script);
        const interpreterDetails = await getInterpreterDetails();
        const pythonPath = interpreterDetails['path'] && interpreterDetails['path'][0];

        cp.exec(`${pythonPath} ${script} ${EXTENSION_ROOT_DIR}`, (error, stdout, stderr) => {
            if (error) {
                console.error(stderr);
                reject(error);
            } else {
                console.log(stdout);
                resolve();
            }
        });
    });
}

callPythonScript('bundled/tool/install_dependencies.py')
```
The reason I was thinking this because it is common for extension/LSP to call tools to perform different actions (think `black` or `ruff` for linting/formating)

**Comment by jitu5** [comment]
> I think it's mainly just a refactoring, you have this now:
> 
> ```ts
> export async function installPythonDependencies(context: any): Promise<void> {
>     return new Promise<void>(async (resolve, reject) => {
>         const script = context.asAbsolutePath('bundled/tool/install_dependencies.py');
>         const interpreterDetails = await getInterpreterDetails();
>         const pythonPath = interpreterDetails['path'] && interpreterDetails['path'][0];
> 
>         cp.exec(`${pythonPath} ${script} ${EXTENSION_ROOT_DIR}`, (error, stdout, stderr) => {
>             if (error) {
>                 console.error(stderr);
>                 reject(error);
>             } else {
>                 console.log(stdout);
>                 resolve();
>             }
>         });
>     });
> }
> ```
> 
> I was thinking something more like this to make it slightly generic. Frankly it is fine to keep it as is as well.
> 
> ```ts
> export async function callPythonScript(context: any, path_to_script: str): Promise<void> {
>     return new Promise<void>(async (resolve, reject) => {
>         const script = context.asAbsolutePath(path_to_script);
>         const interpreterDetails = await getInterpreterDetails();
>         const pythonPath = interpreterDetails['path'] && interpreterDetails['path'][0];
> 
>         cp.exec(`${pythonPath} ${script} ${EXTENSION_ROOT_DIR}`, (error, stdout, stderr) => {
>             if (error) {
>                 console.error(stderr);
>                 reject(error);
>             } else {
>                 console.log(stdout);
>                 resolve();
>             }
>         });
>     });
> }
> 
> callPythonScript('bundled/tool/install_dependencies.py')
> ```
> 
> The reason I was thinking this because it is common for extension/LSP to call tools to perform different actions (think `black` or `ruff` for linting/formating)

@noklam created generic `callPythonScript ` which can be used for other python script run.

**Comment by noklam** [comment]
Is there anything left to be done here? If not I can merge the PR.

**Comment by jitu5** [comment]
> Is there anything left to be done here? If not I can merge the PR.

@noklam Most of the things are done under this PR. we can merge now.

**Pull request merged**